app/src/main/java/com/portfolio/tasky/entry/fragments/LoginFragment.kt

```diff
@@ -1,56 +1,137 @@
 package com.portfolio.tasky.entry.fragments
 
 import android.os.Bundle
+import android.text.Editable
 import android.view.Gravity
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import androidx.databinding.DataBindingUtil
 import androidx.fragment.app.Fragment
-import androidx.transition.Fade
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.viewModels
 import androidx.transition.Slide
-import com.portfolio.tasky.R
+import com.portfolio.tasky.*
 import com.portfolio.tasky.databinding.LayoutLoginBinding
 import com.portfolio.tasky.entry.EntryActivity
-import com.portfolio.tasky.FragmentInflater
-import com.portfolio.tasky.FragmentInflaterImpl
+import com.portfolio.tasky.viewModels.LoginViewModel
+import com.portfolio.tasky.views.TaskyAppCompatEditText
+import dagger.hilt.android.AndroidEntryPoint
 
-class LoginFragment : Fragment(), FragmentInflater by FragmentInflaterImpl() {
+@AndroidEntryPoint
+class LoginFragment : Fragment(), FragmentInflater by FragmentInflaterImpl(), TextChanged  {
     private lateinit var viewBinding: LayoutLoginBinding
 
+    private val viewModel: LoginViewModel by viewModels()
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
     ): View {
-        viewBinding = LayoutLoginBinding.inflate(inflater, container, false)
+        viewBinding = DataBindingUtil.inflate(inflater, R.layout.layout_login, container, false)
+        viewBinding.lifecycleOwner = this
         return viewBinding.root
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
+        setObservers()
         viewBinding.btnLogin.setOnClickListener {
             openRegistrationFragment()
         }
     }
 
+    private fun setObservers() {
+        viewModel.emailChange.observe(viewLifecycleOwner){
+            val emailField = viewBinding.etEmail
+            it?.let { it1 -> emailField.setValid(it1)
+                emailField.setError(emailField.subLayout.etInput.text?.isNotEmpty() as Boolean && !it1)}
+        }
+        viewModel.passwordChange.observe(viewLifecycleOwner){
+            val passwordField = viewBinding.etPassword
+            it?.let { it1 -> passwordField.setValid(it1)
+                passwordField.setError(passwordField.subLayout.etInput.text?.isNotEmpty() as Boolean && !it1)}
+        }
+        viewModel.areFieldsValid.observe(viewLifecycleOwner){
+            it.let {  viewBinding.btnLogin.isEnabled = it == true }
+        }
+    }
+
+    private fun setFieldValidations(loginFragment: LoginFragment?) {
+        val etEmail = viewBinding.etEmail.subLayout.etInput
+        val etPassword = viewBinding.etPassword.subLayout.etInput
+
+        etEmail.addTextChangedListener(loginFragment?.let {
+            TaskyValidationWatcherImpl(viewBinding.etEmail,
+                it
+            )
```

Comment:
Here I prefer calling the viewmodel, and have the viewmodel handle all the validation calls

UIs (Fragments, Activities, etc) should be as dumb as possible. They should just display the state, and on state change, they should call the viewmodel

So maybe create a 
"onEmailChanged" function on the viewmodel. And there, the viewmodel will call the email validator use case

Reply:
I am using a Custom text watcher to listen to text changes, since I have two fields in my layout, I have to check for their ID's in order to activate the correct viewModel.onEmailChange or viewModel.onPasswordChange..

I am not able to comprehend this..please if possible then direct me to some code or video

Reply:
"onEmailChanged" function on the viewmodel. And there, the viewmodel will call the email validator use case


I think this is how it has been implemented..